examples/postgres/advanced-blog-cli/src/auth.rs

```diff
@@ -0,0 +1,164 @@
+use bcrypt::*;
+use diesel::prelude::*;
+use diesel::{self, insert_into};
+use dotenv;
+
+use schema::users;
+
+#[derive(Debug)]
+pub enum AuthenticationError {
+    IncorrectPassword,
+    NoUsernameSet,
+    NoPasswordSet,
+    EnvironmentError(dotenv::Error),
+    BcryptError(BcryptError),
+    DatabaseError(diesel::result::Error),
+}
+
+impl From<BcryptError> for AuthenticationError {
+    fn from(e: BcryptError) -> Self {
+        AuthenticationError::BcryptError(e)
+    }
+}
+
+pub use self::AuthenticationError::{IncorrectPassword, NoPasswordSet, NoUsernameSet};
+
+#[derive(Queryable, Identifiable, Debug, PartialEq)]
+pub struct User {
+    pub id: i32,
+    pub username: String,
+}
+
+#[derive(Queryable)]
+pub struct UserWithPassword {
+    user: User,
+    password: String,
+}
+
+pub fn current_user_from_env(conn: &PgConnection) -> Result<Option<User>, AuthenticationError> {
+    let username = get_username()?;
+    let password = get_password()?;
+    find_user(conn, &username, &password)
+}
+
+pub fn register_user_from_env(conn: &PgConnection) -> Result<User, AuthenticationError> {
+    let username = get_username()?;
+    let password = get_password()?;
+    register_user(conn, &username, &password)
+}
+
+fn find_user(
+    conn: &PgConnection,
+    username: &str,
+    password: &str,
+) -> Result<Option<User>, AuthenticationError> {
+    let user_and_password = users::table
+        .filter(users::username.eq(username))
+        .select(((users::id, users::username), users::hashed_password))
+        .first::<UserWithPassword>(conn)
+        .optional()
+        .map_err(AuthenticationError::DatabaseError)?;
+
+    if let Some(user_and_password) = user_and_password {
+        if verify(password, &user_and_password.password)? {
+            Ok(Some(user_and_password.user))
+        } else {
+            Err(IncorrectPassword)
+        }
+    } else {
+        Ok(None)
+    }
+}
+
+fn register_user(
+    conn: &PgConnection,
+    username: &str,
+    password: &str,
+) -> Result<User, AuthenticationError> {
+    let hashed_password = hash(password, DEFAULT_COST)?;
+    insert_into(users::table)
+        .values((
+            users::username.eq(username),
+            users::hashed_password.eq(hashed_password),
+        ))
+        .returning((users::id, users::username))
+        .get_result(conn)
+        .map_err(AuthenticationError::DatabaseError)
+}
+
+fn get_username() -> Result<String, AuthenticationError> {
+    if_not_present(dotenv::var("BLOG_USERNAME"), NoUsernameSet)
+}
+
+fn get_password() -> Result<String, AuthenticationError> {
+    if_not_present(dotenv::var("BLOG_PASSWORD"), NoPasswordSet)
+}
+
+fn if_not_present<T>(
```

Comment:
That's a weird name… I'd probably write an `fn get_dotenv_or_err_with(&str, AuthenticationError)` that also includes the `dotenv::var`. Making this a very generic function doesn't really help here, as it internally depends on `dotenv::Error` anyway.